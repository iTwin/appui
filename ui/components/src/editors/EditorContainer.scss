/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
.components-editor-container {
  width: 100%;
}