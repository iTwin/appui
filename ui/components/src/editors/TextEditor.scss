/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
@import "@bentley/bwc/lib/mixins";

.components-text-editor {
  font-family:            $bwc-font-family;
  font-size:              $bwc-font-size;
}