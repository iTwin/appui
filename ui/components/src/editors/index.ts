/*---------------------------------------------------------------------------------------------
|  $Copyright: (c) 2018 Bentley Systems, Incorporated. All rights reserved. $
 *--------------------------------------------------------------------------------------------*/
export * from "./EditorContainer";
export * from "./PropertyEditorManager";
export * from "./TextEditor";
