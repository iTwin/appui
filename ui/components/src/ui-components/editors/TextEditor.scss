/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
@import "@bentley/ui-core/lib/ui-core/index";

.components-text-editor {
  min-width:        100%;
  height:           100%;
  margin:           0;
}