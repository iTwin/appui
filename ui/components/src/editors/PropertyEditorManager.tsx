/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module PropertyEditors */

import { PropertyValue } from "../properties/Value";
import { PropertyRecord } from "../properties/Record";
import { PropertyDescription } from "../properties/Description";

import * as React from "react";
import { AsyncValueProcessingResult } from "../converters/TypeConverter";
import { TextEditor } from "./TextEditor";

/** DataControllers can be implemented per typename to validate and commit values.
 */
export interface DataController {
  validateValue(newValue: PropertyValue, record: PropertyRecord): Promise<AsyncValueProcessingResult>;
  commitValue(newValue: PropertyValue, record: PropertyRecord): Promise<AsyncValueProcessingResult>;
}

/** PropertyEditor is the base class for all property editors.
 */
export abstract class PropertyEditorBase implements DataController {
  public customDataController: DataController | undefined = undefined;

  public abstract get reactElement(): React.ReactNode;

  public applyEditorParams(_property: PropertyDescription, _record: PropertyRecord): void { }

  public commitValue(newValue: PropertyValue, record: PropertyRecord): Promise<AsyncValueProcessingResult> {
    if (this.customDataController)
      return this.customDataController.commitValue(newValue, record);

    return Promise.resolve({ encounteredError: false });
  }

  public validateValue(newValue: PropertyValue, record: PropertyRecord): Promise<AsyncValueProcessingResult> {
    if (this.customDataController)
      return this.customDataController.validateValue(newValue, record);

    return Promise.resolve({ encounteredError: false });
  }

}

/** DataControllerBase is the base class for all Data Controllers.
 */
export abstract class DataControllerBase implements DataController {
  public commitValue(_newValue: PropertyValue, _record: PropertyRecord): Promise<AsyncValueProcessingResult> {
    return Promise.resolve({ encounteredError: false });
  }

  public validateValue(_newValue: PropertyValue, _record: PropertyRecord): Promise<AsyncValueProcessingResult> {
    return Promise.resolve({ encounteredError: false });
  }
}

/** Manages Property Editors. Property Editors are registered with and created by the manager.
 */
export class PropertyEditorManager {
  private static _editors: { [index: string]: (any) } = {};
  private static _dataControllers: { [index: string]: (any) } = {};

  public static registerEditor(editType: string, editor: typeof PropertyEditorBase, editorName?: string): void {
    let fullEditorName = editType;
    if (editorName)
      fullEditorName += ":" + editorName;

    if (PropertyEditorManager._editors.hasOwnProperty(fullEditorName)) {
      const nameOfEditor = PropertyEditorManager._editors[fullEditorName].name;
      throw Error("PropertyEditorManager.registerEditor error: type '" + fullEditorName + "' already registered to '" + nameOfEditor + "'");
    }
    PropertyEditorManager._editors[fullEditorName] = editor;
  }

  public static registerDataController(controllerName: string, controller: typeof DataControllerBase): void {
    if (PropertyEditorManager._dataControllers.hasOwnProperty(controllerName)) {
      throw Error("PropertyEditorManager.RegisterDataController error: type '" + controllerName + "' already registered to '" + (typeof PropertyEditorManager._dataControllers[controllerName]).toString() + "'");
    }
    PropertyEditorManager._dataControllers[controllerName] = controller;
  }

  public static createEditor(editType: string, editorName?: string, dataContollerName?: string): PropertyEditorBase | null {
    let fullEditorName = editType;
    if (editorName)
      fullEditorName += ":" + editorName;

    let editor: PropertyEditorBase;
    if (PropertyEditorManager._editors.hasOwnProperty(fullEditorName))
      editor = new PropertyEditorManager._editors[fullEditorName]();
    else if (PropertyEditorManager._editors.hasOwnProperty(editType))
      editor = new PropertyEditorManager._editors[editType]();
    else
      editor = new BasicPropertyEditor();

    if (dataContollerName) {
      if (PropertyEditorManager._dataControllers.hasOwnProperty(dataContollerName))
        editor.customDataController = new PropertyEditorManager._dataControllers[dataContollerName]();
    }

    return editor;
  }

  public static hasCustomEditor(editType: string, editorName: string): boolean {
    const fullEditorName = editType + ":" + editorName;
    return PropertyEditorManager._editors.hasOwnProperty(fullEditorName);
  }
}

/** BasicPropertyEditor React component that uses the [[TextEditor]] property editor. */
export class BasicPropertyEditor extends PropertyEditorBase {

  public get reactElement(): React.ReactNode {
    return <TextEditor />;
  }
}

PropertyEditorManager.registerEditor("text", BasicPropertyEditor);
PropertyEditorManager.registerEditor("string", BasicPropertyEditor);
