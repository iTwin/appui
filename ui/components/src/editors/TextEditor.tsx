/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module PropertyEditors */

import * as React from "react";
import classnames from "classnames";
import { PropertyRecord } from "../properties/Record";
import { PropertyValueFormat, PrimitiveValue } from "../properties/Value";
import { TypeConverterManager } from "../converters/TypeConverterManager";

import "./TextEditor.scss";

/** Properties for [[TextEditor]] component */
export interface TextEditorProps {
  onBlur?: (event: any) => void;
  value?: PropertyRecord;
}

interface TextEditorState {
  inputValue: string;
}

/** TextEditor React component that is a property editor with text input  */
export class TextEditor extends React.Component<TextEditorProps, TextEditorState> {
  private _input: HTMLInputElement | null = null;
  private _isMounted = false;

  /** @hidden */
  public readonly state: Readonly<TextEditorState> = {
    inputValue: "",
  };

  public getValue(): string {
    return this.state.inputValue;
  }

  public getInputNode(): HTMLInputElement | null {
    return this._input;
  }

  private _updateInputValue = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (this._isMounted)
      this.setState({
        inputValue: e.target.value,
      });
  }

  public componentDidMount() {
    this._isMounted = true;
    this.getInitialValue(); // tslint:disable-line:no-floating-promises
  }

  public componentWillUnmount() {
    this._isMounted = false;
  }

  private async getInitialValue() {
    const record = this.props.value;
    let initialValue = "";

    if (record && record.value.valueFormat === PropertyValueFormat.Primitive) {
      const value = (record.value as PrimitiveValue).value;
      initialValue = await TypeConverterManager.getConverter(record.property.typename).convertPropertyToString(record.property, value);
    }

    if (this._isMounted)
      this.setState(
        () => ({ inputValue: initialValue }),
        () => {
          if (this._input) {
            this._input.focus();
            this._input.select();
          }
        },
      );
  }

  public render() {
    const className = classnames("cell", "components-cell-editor", "components-text-editor");

    return (
      <input
        ref={(node) => this._input = node}
        type="text"
        onBlur={this.props.onBlur}
        className={className}
        defaultValue={this.state.inputValue}
        onChange={this._updateInputValue}
        data-testid="components-text-editor"
      />
    );
  }
}
