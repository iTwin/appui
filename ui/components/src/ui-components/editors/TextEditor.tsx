/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module PropertyEditors */

import * as React from "react";
import classnames from "classnames";
import { PropertyValueFormat, PropertyValue, PrimitiveValue, PropertyEditorParams, PropertyEditorParamTypes, InputEditorSizeParams } from "@bentley/imodeljs-frontend";
import { PropertyEditorProps, TypeEditor } from "./EditorContainer";
import { TypeConverterManager } from "../converters/TypeConverterManager";

import "./TextEditor.scss";

/** @internal */
interface TextEditorState {
  inputValue: string;
  readonly: boolean;
  isDisabled?: boolean;
  size?: number;
  maxLength?: number;
}

/** TextEditor React component that is a property editor with text input
 * @beta
 */
export class TextEditor extends React.PureComponent<PropertyEditorProps, TextEditorState> implements TypeEditor {
  private _input: HTMLInputElement | null = null;
  private _isMounted = false;

  /** @internal */
  public readonly state: Readonly<TextEditorState> = {
    inputValue: "",
    readonly: false,
  };

  public getValue(): string {
    return this.state.inputValue;
  }

  public async getPropertyValue(): Promise<PropertyValue | undefined> {
    const record = this.props.propertyRecord;
    let propertyValue: PropertyValue | undefined;

    // istanbul ignore else
    if (record && record.value.valueFormat === PropertyValueFormat.Primitive) {
      propertyValue = await TypeConverterManager.getConverter(record.property.typename).convertFromStringToPropertyValue(this.state.inputValue, record);
      (propertyValue as PrimitiveValue).displayValue = this.state.inputValue;
    }

    return propertyValue;
  }

  private setFocus(): void {
    // istanbul ignore else
    if (this._input && !this.state.isDisabled) {
      this._input.focus();
    }
  }

  private _updateInputValue = (e: React.ChangeEvent<HTMLInputElement>) => {
    // istanbul ignore else
    if (this._isMounted)
      this.setState({
        inputValue: e.target.value,
      });
  }

  /** @internal */
  public componentDidMount() {
    this._isMounted = true;
    this.setStateFromProps(); // tslint:disable-line:no-floating-promises
  }

  /** @internal */
  public componentWillUnmount() {
    this._isMounted = false;
  }

  /** @internal */
  public componentDidUpdate(prevProps: PropertyEditorProps) {
    if (this.props.propertyRecord !== prevProps.propertyRecord) {
      this.setStateFromProps(); // tslint:disable-line:no-floating-promises
    }
  }

  private async setStateFromProps() {
    const record = this.props.propertyRecord;
    let initialValue = "";

    // istanbul ignore else
    if (record && record.value.valueFormat === PropertyValueFormat.Primitive) {
      const value = (record.value as PrimitiveValue).value;
      initialValue = await TypeConverterManager.getConverter(record.property.typename).convertPropertyToString(record.property, value);
    }

    const readonly = record && undefined !== record.isReadonly ? record.isReadonly : false;
    let size: number | undefined;
    let maxLength: number | undefined;

    const isDisabled = record ? record.isDisabled : undefined;

    if (record && record.property && record.property.editor && record.property.editor.params) {
      const editorSizeParams = record.property.editor.params.find((param: PropertyEditorParams) => param.type === PropertyEditorParamTypes.InputEditorSize) as InputEditorSizeParams;
      // istanbul ignore else
      if (editorSizeParams) {
        // istanbul ignore else
        if (editorSizeParams.size)
          size = editorSizeParams.size;
        // istanbul ignore else
        if (editorSizeParams.maxLength)
          maxLength = editorSizeParams.maxLength;
      }
    }

    // istanbul ignore else
    if (this._isMounted)
      this.setState(
        { inputValue: initialValue, readonly, size, maxLength, isDisabled },
        () => {
          if (this.props.setFocus) {
            this.setFocus();
            // istanbul ignore else
            if (this._input)
              this._input.select();
          }
        },
      );
  }

  /** @internal */
  public render() {
    const className = classnames("cell", "components-cell-editor", "components-text-editor", this.props.className);

    return (
      <input
        ref={(node) => this._input = node}
        type="text"
        onBlur={this.props.onBlur}
        className={className}
        style={this.props.style}
        readOnly={this.state.readonly}
        disabled={this.state.isDisabled}
        size={this.state.size}
        maxLength={this.state.maxLength}
        defaultValue={this.state.inputValue}
        onChange={this._updateInputValue}
        data-testid="components-text-editor"
      />
    );
  }
}
