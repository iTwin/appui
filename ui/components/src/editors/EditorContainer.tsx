/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module PropertyEditors */

import * as React from "react";
import { PropertyRecord } from "../properties/Record";
import { PropertyEditorBase, PropertyEditorManager } from "./PropertyEditorManager";

/** Arguments for the Property Updated event callback */
export interface PropertyUpdatedArgs {
  /** The property being updated. */
  propertyRecord?: PropertyRecord;
  /** The new value for the property. */
  newValue: any;
}

/** [[EditorContainer]] React component properties */
export interface EditorContainerProps {
  propertyRecord?: PropertyRecord;
  title: string;
  onCommit: (commit: PropertyUpdatedArgs) => void;
  onCommitCancel: () => void;
}

/**
 * EditorContainer React component
 */
export class EditorContainer extends React.Component<EditorContainerProps> {

  private _editorRef: any;
  private _propertyEditor: PropertyEditorBase | null = null;

  private getEditor(): any {
    return this._editorRef;
  }

  private createEditor(): React.ReactNode {
    const editorRef = (c: any) => this._editorRef = c;

    const editorProps = {
      ref: editorRef,
      onBlur: this._commit,
      value: this.props.propertyRecord,
    };

    if (this.props.propertyRecord) {
      let editorNode: React.ReactNode;
      const propDescription = this.props.propertyRecord.property;

      if (propDescription.typename) {
        const editorName = propDescription.editor !== undefined ? propDescription.editor.name : undefined;
        this._propertyEditor = PropertyEditorManager.createEditor(propDescription.typename, editorName, propDescription.dataController);
        if (this._propertyEditor)
          editorNode = this._propertyEditor.reactElement;
      }

      if (React.isValidElement(editorNode)) {
        return React.cloneElement(editorNode, editorProps);
      }
    }

    return null;
  }

  private _handleBlur = (e: React.FocusEvent) => {
    e.stopPropagation();
  }

  private _handleClick = (e: React.MouseEvent) => {
    e.stopPropagation();
  }

  private _handleKeyDown = (e: React.KeyboardEvent) => {
    switch (e.key) {
      case "Escape":
        this.onPressEscape(e);
        break;
      case "Enter":
        this.onPressEnter(e);
        break;
      case "Tab":
        this.onPressTab(e);
        break;
    }

    // Prevent the arrow keys from bubbling up to the ReactDataGrid
    if (e.keyCode >= 37 && e.keyCode <= 40)
      e.stopPropagation();
  }

  private _handleRightClick = (e: React.MouseEvent) => {
    e.stopPropagation();
  }

  private onPressEscape(e: React.KeyboardEvent): void {
    if (!this.editorIsSelectOpen()) {
      this._commitCancel();
    } else {
      // prevent event from bubbling if editor has results to select
      e.stopPropagation();
    }
  }

  private onPressEnter(e: React.KeyboardEvent): void {
    this._commit({ key: e.key });
  }

  private onPressTab(e: React.KeyboardEvent): void {
    this._commit({ key: e.key });
  }

  private editorIsSelectOpen(): boolean {
    if (isFunction(this.getEditor().isSelectOpen)) {
      return this.getEditor().isSelectOpen();
    }

    return false;
  }

  private async isNewValueValid(value: any): Promise<boolean> {
    if (isFunction(this.getEditor().validate)) {
      const isValid = this.getEditor().validate(value);
      if (!isValid) {
        this.setState({ isInvalid: !isValid });
        return isValid;
      }
    }

    if (this._propertyEditor && this.props.propertyRecord) {
      const valueResult = await this._propertyEditor.validateValue(value, this.props.propertyRecord);
      if (valueResult.encounteredError) {
        this.setState({ isInvalid: valueResult.encounteredError });
        // TODO - display InputField
        return !valueResult.encounteredError;
      }
    }

    return true;
  }

  private _commit = (_args: { key: string }) => {
    const newValue = this.getEditor().getValue();
    if (this.isNewValueValid(newValue)) {
      this.props.onCommit({ propertyRecord: this.props.propertyRecord, newValue });
    }
  }

  private _commitCancel = () => {
    this.props.onCommitCancel();
  }

  public componentDidMount() {
    const inputNode = this.getInputNode();
    if (inputNode) {
      inputNode.focus();
    }
  }

  private getInputNode(): HTMLInputElement | null {
    if (this.getEditor() && isFunction(this.getEditor().getInputNode))
      return this.getEditor().getInputNode();
    return null;
  }

  public render() {
    return (
      <div
        onBlur={this._handleBlur}
        onKeyDown={this._handleKeyDown}
        onClick={this._handleClick}
        onContextMenu={this._handleRightClick}
        title={this.props.title}
      >
        {this.createEditor()}
      </div>
    );
  }
}

const isFunction = (functionToCheck: any): boolean => {
  const getType = {};
  return functionToCheck && getType.toString.call(functionToCheck) === "[object Function]";
};
