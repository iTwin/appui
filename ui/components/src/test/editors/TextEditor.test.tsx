/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/

import * as React from "react";
import { mount, shallow } from "enzyme";
import { expect } from "chai";
import { TextEditor } from "../../ui-components/editors/TextEditor";
import TestUtils from "../TestUtils";
import { PropertyEditorParamTypes, InputEditorSizeParams, IconEditorParams, PropertyEditorInfo } from "@bentley/imodeljs-frontend";

describe("<TextEditor />", () => {
  it("should render", () => {
    mount(<TextEditor />);
  });

  it("renders correctly", () => {
    shallow(<TextEditor />).should.matchSnapshot();
  });

  it("getValue returns proper value after componentDidMount & setState", async () => {
    const record = TestUtils.createPrimitiveStringProperty("Test", "MyValue");
    const wrapper = mount(<TextEditor propertyRecord={record} />);

    await TestUtils.flushAsyncOperations();
    const textEditor = wrapper.instance() as TextEditor;
    expect(textEditor.getValue()).to.equal("MyValue");

    wrapper.unmount();
  });

  it("HTML input onChange updates value", () => {
    const record = TestUtils.createPrimitiveStringProperty("Test1", "MyValue");
    const wrapper = mount(<TextEditor propertyRecord={record} />);
    const textEditor = wrapper.instance() as TextEditor;
    const inputNode = wrapper.find("input");

    expect(inputNode.length).to.eq(1);
    if (inputNode) {
      const testValue = "My new value";
      inputNode.simulate("change", { target: { value: testValue } });
      wrapper.update();
      expect(textEditor.getValue()).to.equal(testValue);
    }
  });

  it("componentDidUpdate updates the value", async () => {
    const record = TestUtils.createPrimitiveStringProperty("Test", "MyValue");
    const wrapper = mount(<TextEditor propertyRecord={record} />);

    await TestUtils.flushAsyncOperations();
    const textEditor = wrapper.instance() as TextEditor;
    expect(textEditor.getValue()).to.equal("MyValue");

    const testValue = "MyNewValue";
    const newRecord = TestUtils.createPrimitiveStringProperty("Test", testValue);
    wrapper.setProps({ propertyRecord: newRecord });
    await TestUtils.flushAsyncOperations();
    expect(textEditor.getValue()).to.equal(testValue);

    wrapper.unmount();
  });

  it("should support InputEditorSize params", async () => {
    const size = 4;
    const maxLength = 60;
    const editorInfo: PropertyEditorInfo = {
      params: [
        {
          type: PropertyEditorParamTypes.InputEditorSize,
          size,
          maxLength,
        } as InputEditorSizeParams,
      ],
    };

    const record = TestUtils.createPrimitiveStringProperty("Test", "MyValue", "Test", editorInfo);
    const wrapper = mount(<TextEditor propertyRecord={record} />);
    await TestUtils.flushAsyncOperations();

    const textEditor = wrapper.find(TextEditor);
    expect(textEditor.length).to.eq(1);
    expect(textEditor.state("size")).to.eq(size);
    expect(textEditor.state("maxLength")).to.eq(maxLength);

    wrapper.unmount();
  });

  it("should support IconEditor params", async () => {
    const iconSpec = "icon-placeholder";
    const editorInfo: PropertyEditorInfo = {
      params: [
        {
          type: PropertyEditorParamTypes.Icon,
          definition: { iconSpec },
        } as IconEditorParams,
      ],
    };

    const record = TestUtils.createPrimitiveStringProperty("Test", "MyValue", "Test", editorInfo);
    const wrapper = mount(<TextEditor propertyRecord={record} />);
    await TestUtils.flushAsyncOperations();

    const textEditor = wrapper.find(TextEditor);
    expect(textEditor.length).to.eq(1);
    expect(textEditor.state("iconSpec")).to.eq(iconSpec);

    wrapper.unmount();
  });
});
