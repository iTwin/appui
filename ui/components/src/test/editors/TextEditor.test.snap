// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextEditor /> renders correctly 1`] = `
<Input
  className="components-cell-editor components-text-editor"
  data-testid="components-text-editor"
  defaultValue=""
  onChange={[Function]}
  readOnly={false}
  style={
    Object {
      "minWidth": "6em",
    }
  }
  type="text"
/>
`;
